stan/math/prim/prob/wiener4_lcdf.hpp

```diff
@@ -0,0 +1,913 @@
+    const auto exp_minus_two_av = exp(minus_two_av);
+    if ((exp_minus_two_va_one_minus_w >= nearly_one)
+        || (exp_minus_two_av >= nearly_one)) {
+      return ret_t(-w);
+    }
+    prob = LOG_TWO - log1p(-exp_minus_two_va_one_minus_w);
+    ret_t log_quotient;
+    if (minus_two_va_one_minus_w > minus_two_av) {
+      log_quotient = log_diff_exp(minus_two_va_one_minus_w, minus_two_av)
+                     - log1p(-exp_minus_two_av);
+    } else if (minus_two_va_one_minus_w < minus_two_av) {
+      log_quotient = log_diff_exp(minus_two_av, minus_two_va_one_minus_w)
+                     - log1p(-exp_minus_two_av);
+    } else {
+      log_quotient = NEGATIVE_INFTY - log1p(-exp_minus_two_av);
+    }
+    if (log(w) > log_quotient) {
+      prob += log_diff_exp(log(w), log_quotient);
+      prob = -exp(prob);
+    } else {
+      prob += log_diff_exp(log_quotient, log(w));
+      prob = exp(prob);
+    }
+  }
+  return (is_inf(prob)) ? NEGATIVE_INFTY : prob;
+}
+
+/**
+ * Calculate the probability term 'P' on log scale
+ *
+ * @param a The boundary separation
+ * @param v The drift rate
+ * @param w The relative starting point
+ * @return 'P' term
+ */
+template <typename T_a, typename T_w, typename T_v>
+inline auto log_probability_GradW(const T_a& a, const T_v& v,
+                                  const T_w& w) noexcept {
+  using ret_t = return_type_t<T_a, T_w, T_v>;
+  auto nearly_one = ret_t(1.0 - 1.0e-6);
+  if (fabs(v) == 0.0) {
+    return ret_t(-1 / (1.0 - w));
+  }
+  const auto sign_v = (v < 0) ? 1 : -1;
+  const auto sign_two_va_one_minus_w = sign_v * (2.0 * v * a * (1.0 - w));
+  const auto exp_arg = exp(sign_two_va_one_minus_w);
+  if (exp_arg >= nearly_one) {
+    return ret_t(-1 / (1.0 - w));
+  }
+  auto prob = LOG_TWO + log(fabs(v)) + log(a) - log1p(-exp_arg);
+  prob = (v < 0) ? prob + sign_two_va_one_minus_w : prob;
+  return -exp(prob);
+}
+
+/**
+ * Calculate 'K_small', the number of terms needed for small y
+ *
+ * @tparam Distribution Whether the calculation is for the distribution
+ * @tparam GradA Whether the calculation is for gradient w.r.t. 'a'
+ * @tparam GradV Whether the calculation is for gradient w.r.t. 'v'
+ * @tparam GradW Whether the calculation is for gradient w.r.t. 'w'
+ *
+ * @param y A scalar variable; the reaction time in seconds
+ * @param a The boundary separation
+ * @param v The drift rate
+ * @param w The relative starting point
+ * @param precision The error tolerance
+ * @return 'K_small' term
+ */
+template <GradientCalc Distribution = GradientCalc::OFF,
+          GradientCalc GradA = GradientCalc::OFF,
+          GradientCalc GradV = GradientCalc::OFF,
+          GradientCalc GradW = GradientCalc::OFF, typename T_y, typename T_a,
+          typename T_v, typename T_w, typename T_prec>
+inline auto K_Small(const T_y& y, const T_a& a, const T_v& v, const T_w& w,
+                    T_prec&& precision) noexcept {
```

Comment:
Are any of the `GradientCalc`'s ever true simultaneously. If not then I think this should be broken up into multiple functions or if they are only called once just inlined where they are called.

For reference, when we use template `bool` values like `if (GradA)` the compiler knows at compile time whether that value is true or false. That means the compiler can just eliminate the code inside of the if during compilation during a process known as Dead Code Elimination (DCE).

We the compile time bool and DCE in in the distributions because there are a lot of common operations between the derivatives. Here it seems like there is much less.

For places where functions are used once can we just inline them? And for functions like this that only have on one bool true at a time, and are used multiple times with each bool flag, I think it would be good to break them up into multiple functions

Reply:
Ok. I reduced and inlined some functions.